src/cmd/ksh93/bltins/misc.c

```diff
@@ -298,7 +298,8 @@ int b_dot_cmd(int n, char *argv[], Shbltin_t *context) {
     }
     sh_popcontext(shp, &buff);
     if (buffer) free(buffer);
-    if (!np) free(shp->st.filename);
+    // Following line was commented as a workaround for https://github.com/att/ast/issues/582
+    // if (!np) free(shp->st.filename);
```

Comment:
As best I can tell all but one of the assignments to `shp->st.filename` involve `strdup()` or equivalent. So this introduces a memory leak. The one place I could find which does not make a copy that can be freed is in *src/cmd/ksh93/sh/xec.c*:

```
shp->st.filename = fp->node->nvalue.rp->fname;
```

Does wrapping that in `strdup()` while leaving this `free()` in effect fix it?

Reply:
Thanks for the hint. It was not this statement, but the one I fixed in https://github.com/att/ast/pull/592/commits/b089e63248c43b1326bedda6f7d454f4497418a5 that fixes the crash.